File: crdt_sync_dart/lib/src/hlc.dart
/// A Hybrid Logical Clock: a totally-ordered, monotonic per-node timestamp.
///
/// Combines wall-clock time with a logical counter so two ticks issued by
/// the same node in the same millisecond still get a strict order, and two
/// ticks from different nodes are always comparable (the node id breaks
/// ties), without requiring synchronized clocks across devices. Mirrors
/// `crdt_sync._hlc.Hlc` on the Python side field-for-field so the two
/// serialize to and parse the same wire format.
class Hlc implements Comparable<Hlc> {
  const Hlc({
    required this.wallTimeMs,
    required this.counter,
    required this.nodeId,
  });

  final int wallTimeMs;
  final int counter;
  final String nodeId;

  static const _isoPrefixLen = 23; // "YYYY-MM-DDTHH:MM:SS.mmm"

  /// Returns the next clock value for [nodeId].
  ///
  /// Passing [previous] (this node's last-issued clock) is what makes the
  /// clock monotonic even when the wall clock hasn't advanced (or has gone
  /// backwards). [wallTimeMsOverride] is exposed for deterministic tests.
  static Hlc newTick(String nodeId, {Hlc? previous, int? wallTimeMsOverride}) {
    final now =
        wallTimeMsOverride ?? DateTime.now().toUtc().millisecondsSinceEpoch;
    if (previous == null) {
      return Hlc(wallTimeMs: now, counter: 0, nodeId: nodeId);
    }
    final newWall = now > previous.wallTimeMs ? now : previous.wallTimeMs;
    final counter = newWall == previous.wallTimeMs ? previous.counter + 1 : 0;
    return Hlc(wallTimeMs: newWall, counter: counter, nodeId: nodeId);
  }

  /// Serializes to a lexicographically sortable string:
  /// `<iso8601-millis>Z-<counter:hex4>-<nodeId>`.
  String toStr() {
    final dt = DateTime.fromMillisecondsSinceEpoch(wallTimeMs, isUtc: true);
    String two(int n) => n.toString().padLeft(2, '0');
    final iso =
        '${dt.year.toString().padLeft(4, '0')}-${two(dt.month)}-${two(dt.day)}'
        'T${two(dt.hour)}:${two(dt.minute)}:${two(dt.second)}'
        '.${(wallTimeMs % 1000).toString().padLeft(3, '0')}';
    final counterHex = counter.toRadixString(16).padLeft(4, '0');
    return '${iso}Z-$counterHex-$nodeId';
  }

  /// Parses the format produced by [toStr]. Throws [FormatException].
  static Hlc fromStr(String text) {
    final zIndex = text.indexOf('Z-');
    if (zIndex == -1 || zIndex != _isoPrefixLen) {
      throw FormatException('not a valid Hlc string', text);
    }
    final isoPart = text.substring(0, zIndex);
    final rest = text.substring(zIndex + 2);
    final dashIndex = rest.indexOf('-');
    if (dashIndex == -1) {
      throw FormatException('not a valid Hlc string', text);
    }
    final counterHex = rest.substring(0, dashIndex);
    final nodeId = rest.substring(dashIndex + 1);
    final dt = DateTime.parse('${isoPart}Z');
    return Hlc(
      wallTimeMs: dt.millisecondsSinceEpoch,
      counter: int.parse(counterHex, radix: 16),
      nodeId: nodeId,
    );
  }

  @override
  int compareTo(Hlc other) {
    if (wallTimeMs != other.wallTimeMs) {
      return wallTimeMs.compareTo(other.wallTimeMs);
    }
    if (counter != other.counter) {
      return counter.compareTo(other.counter);
    }
    return nodeId.compareTo(other.nodeId);
  }

  bool operator >(Hlc other) => compareTo(other) > 0;
  bool operator >=(Hlc other) => compareTo(other) >= 0;
  bool operator <(Hlc other) => compareTo(other) < 0;
  bool operator <=(Hlc other) => compareTo(other) <= 0;

  @override
  bool operator ==(Object other) =>
      other is Hlc &&
      wallTimeMs == other.wallTimeMs &&
      counter == other.counter &&
      nodeId == other.nodeId;

  @override
  int get hashCode => Object.hash(wallTimeMs, counter, nodeId);

  @override
  String toString() =>
      'Hlc(wallTimeMs: $wallTimeMs, counter: $counter, nodeId: $nodeId)';
}

File: crdt-sync/crdt_sync/_hlc.py
"""Hybrid Logical Clock: a totally-ordered, monotonic per-node timestamp.

An HLC combines wall-clock time with a logical counter so that two ticks
issued by the same node in the same millisecond still get a strict order,
and two ticks from different nodes are always comparable (the node id breaks
ties), without requiring synchronized clocks across devices.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
import time

_ISO_PREFIX_LEN = 23  # len("YYYY-MM-DDTHH:MM:SS.mmm")
_EPOCH = datetime(1970, 1, 1, tzinfo=UTC)


@dataclass(frozen=True, order=True)
class Hlc:
    """A Hybrid Logical Clock value.

    Ordered by ``(wall_time_ms, counter, node_id)`` tuple comparison, which
    is also the order ``@dataclass(order=True)`` derives from the field
    declaration order below -- two clocks from different nodes never compare
    equal, since ``node_id`` is the final, always-distinguishing tiebreaker.
    """

    wall_time_ms: int
    counter: int
    node_id: str

    @staticmethod
    def new_tick(
        node_id: str,
        previous: Hlc | None = None,
        wall_time_ms: int | None = None,
    ) -> Hlc:
        """Return the next clock value for ``node_id``.

        Args:
            node_id: The issuing node's identifier.
            previous: This node's last-issued clock value, if any. Passing
                the previous value is what makes the clock monotonic even
                when the wall clock hasn't advanced (or has gone backwards).
            wall_time_ms: Override for the current wall-clock time in
                milliseconds since the Unix epoch; defaults to the real
                clock. Exposed for deterministic tests.

        Returns:
            A new ``Hlc`` strictly greater than ``previous`` (if given).
        """
        now = wall_time_ms if wall_time_ms is not None else int(time.time() * 1000)
        if previous is None:
            return Hlc(wall_time_ms=now, counter=0, node_id=node_id)
        new_wall = max(now, previous.wall_time_ms)
        counter = previous.counter + 1 if new_wall == previous.wall_time_ms else 0
        return Hlc(wall_time_ms=new_wall, counter=counter, node_id=node_id)

    def to_str(self) -> str:
        """Serialize to a human-readable, lexicographically sortable string.

        Format: ``<iso8601-millis>Z-<counter:04x>-<node_id>``. The iso8601
        portion is fixed-width (always exactly ``YYYY-MM-DDTHH:MM:SS.mmm``),
        which is what makes plain string comparison agree with ``Hlc``
        comparison for the wall-clock component.
        """
        seconds, millis = divmod(self.wall_time_ms, 1000)
        dt = datetime.fromtimestamp(seconds, tz=UTC)
        iso = dt.strftime("%Y-%m-%dT%H:%M:%S")
        return f"{iso}.{millis:03d}Z-{self.counter:04x}-{self.node_id}"

    @classmethod
    def from_str(cls, text: str) -> Hlc:
        """Parse the format produced by :meth:`to_str`.

        Raises:
            ValueError: If ``text`` is not in the expected format.
        """
        iso_part, sep, rest = text.partition("Z-")
        if not sep or len(iso_part) != _ISO_PREFIX_LEN:
            msg = f"not a valid Hlc string: {text!r}"
            raise ValueError(msg)
        counter_hex, sep, node_id = rest.partition("-")
        if not sep:
            msg = f"not a valid Hlc string: {text!r}"
            raise ValueError(msg)
        dt = datetime.strptime(iso_part, "%Y-%m-%dT%H:%M:%S.%f").replace(tzinfo=UTC)
        delta = dt - _EPOCH
        wall_time_ms = (
            delta.days * 86_400_000 + delta.seconds * 1000 + delta.microseconds // 1000
        )
        return cls(
            wall_time_ms=wall_time_ms, counter=int(counter_hex, 16), node_id=node_id
        )
